mindmeld/components/question_answerer.py

```diff
@@ -2741,11 +2789,122 @@ def is_vector_field(self):
             return self.type in self.VECTOR_TYPES
 
 
+class QuestionAnswererFactory:
+    """
+    Factory class for creating QuestionAnswerers
+
+    usage
+        >>> question_answerer = QuestionAnswererFactory.create_question_answerer(**kwargs)
+        >>> question_answerer.load_kb(...)
+        >>> question_answerer.get(...) # .get(...) or .build_search(...)
+    """
+
+    @classmethod
+    def create_question_answerer(cls, **kwargs):
+        """
+        Args:
+            app_path (str, optional): The path to the directory containing the app's data. If
+                provided, used to obtain default 'app_namespace' and QA configurations
+            app_namespace (str, optional): The namespace of the app. Used to prevent
+                collisions between the indices of this app and those of other apps.
+            config (dict, optional): The QA config if passed directly rather than loaded from the
+                app config
+            resource_loader (ResourceLoader, optional): An object which can load resources for the
+                question answerer.
+        """
+
+        config = cls._get_config(kwargs.get("config"), kwargs.get("app_path"))
+        config = cls._correct_deprecated_qa_config(config)
+        kwargs.update({"config": config})
+        return cls._get_question_answerer_class(config.get("model_type"))(**kwargs)
+
+    @staticmethod
+    def _get_config(config=None, app_path=None):
+        if not config:
+            return get_classifier_config("question_answering", app_path=app_path)
+        return config
+
+    @staticmethod
+    def _correct_deprecated_qa_config(config):
+        """
+        for backwards compatability
+          if the config is supplied in deprecated format, its format is corrected and returned,
+          else it is not modified and returned as-is
+
+        deprecated usage
+            >>> config = {
+                    "model_type": "keyword",  # or "text", "embedder", "embedder_keyword", etc.
+                    "model_settings": {
+                        ...
+                    }
+                }
+
+        new usage
+            >>> config = {
+                    "model_type": "elasticsearch",  # or "native"
+                    "model_settings": {
+                        "query_type": "keyword",  # or "text", "embedder", "embedder_keyword", etc.
+                        ...
+                    }
+                }
+        """
+
+        if not config.get("model_settings", {}).get("query_type"):
+            model_type = config.get("model_type")
+            if not model_type:
+                msg = f"Invalid 'model_type': {model_type} found while creating a QuestionAnswerer"
+                raise ValueError(msg)
+            if model_type in QUESTION_ANSWERER_MODEL_MAPPINGS:
+                raise ValueError(
+                    "Could not find `query_type` in `model_settings` of question answerer")
```

Comment:
emit the whole config here as well

Reply:
the `if model_type in QUESTION_ANSWERER_MODEL_MAPPINGS:` conditional is for scenario wherein we anticipated an older config version's `model_type` values (because we did not see `query_type` key which was not available in older config version), but we observed newer config version's `model_type`. This is an edge case when users mix up the versions.